File: nodejs/src/cert/cert.js
'use strict';

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

function ensureCerts(certDir) {
  fs.mkdirSync(certDir, { recursive: true });
  const certFile = path.join(certDir, 'server.crt');
  const keyFile = path.join(certDir, 'server.key');

  if (fs.existsSync(certFile) && fs.existsSync(keyFile)) {
    return { certFile, keyFile };
  }

  try {
    execSync(
      `openssl req -x509 -newkey rsa:2048 -keyout "${keyFile}" -out "${certFile}" -days 3650 -nodes -subj "/CN=localhost/OU=Dev/O=MockServer/L=City/ST=State/C=US"`,
      { stdio: 'pipe' }
    );
    console.log(`[Certs] Self-signed certificate created at ${certDir}`);
  } catch (e) {
    throw new Error(`Could not generate certificates (openssl required): ${e.message}`);
  }

  return { certFile, keyFile };
}

module.exports = { ensureCerts };

File: go/internal/cert/cert.go
// Package cert generates self-signed TLS certificates for local HTTPS/WSS.
package cert

import (
	"crypto/rand"
	"crypto/rsa"
	"crypto/x509"
	"crypto/x509/pkix"
	"encoding/pem"
	"fmt"
	"math/big"
	"os"
	"path/filepath"
	"time"
)

// Ensure returns paths to cert and key files, generating them if needed.
func Ensure(dir string) (certFile, keyFile string, err error) {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return "", "", err
	}
	certFile = filepath.Join(dir, "server.crt")
	keyFile = filepath.Join(dir, "server.key")
	if fileExists(certFile) && fileExists(keyFile) {
		return certFile, keyFile, nil
	}
	return certFile, keyFile, generate(certFile, keyFile)
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func generate(certFile, keyFile string) error {
	key, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		return err
	}
	serial, _ := rand.Int(rand.Reader, new(big.Int).Lsh(big.NewInt(1), 128))
	template := x509.Certificate{
		SerialNumber: serial,
		Subject: pkix.Name{
			CommonName:   "localhost",
			Organization: []string{"MockServer"},
		},
		NotBefore:             time.Now(),
		NotAfter:              time.Now().Add(3650 * 24 * time.Hour),
		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
		BasicConstraintsValid: true,
		DNSNames:              []string{"localhost"},
	}
	der, err := x509.CreateCertificate(rand.Reader, &template, &template, &key.PublicKey, key)
	if err != nil {
		return err
	}
	certOut, err := os.Create(certFile)
	if err != nil {
		return err
	}
	defer certOut.Close()
	if err := pem.Encode(certOut, &pem.Block{Type: "CERTIFICATE", Bytes: der}); err != nil {
		return err
	}
	keyOut, err := os.Create(keyFile)
	if err != nil {
		return err
	}
	defer keyOut.Close()
	if err := pem.Encode(keyOut, &pem.Block{Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(key)}); err != nil {
		return err
	}
	fmt.Printf("[Certs] Self-signed certificate created at %s\n", filepath.Dir(certFile))
	return nil
}
